File: asme/ops/serializers/work_orders.py
"""JSON shapes for work orders (see the plan's "Serializer shapes").

``work_order`` is the list/item shape; ``work_order_detail`` adds status
history, time and cost entries, dependencies, children and related assets.
``serialize_many`` batches the per-row sub-work-order counts so a list page is
two queries instead of one per row.
"""

from __future__ import annotations

from sqlalchemy import case, func, select

from asme.extensions import db
from asme.ops.models import WorkOrder
from asme.ops.serializers import asset_ref, category_ref, iso, location_ref, money, project_ref, team_ref, uid, user_ref, vendor_ref


def sub_counts_for(work_order_ids) -> dict:
    """``{parent_id: {"total": n, "done": n}}`` for the given parent ids in one query."""
    ids = [i for i in work_order_ids if i is not None]
    if not ids:
        return {}
    rows = db.session.execute(
        select(
            WorkOrder.parent_work_order_id,
            func.count(WorkOrder.id),
            func.sum(case((WorkOrder.status == "done", 1), else_=0)),
        )
        .where(WorkOrder.parent_work_order_id.in_(ids))
        .group_by(WorkOrder.parent_work_order_id)
    ).all()
    return {parent_id: {"total": int(total), "done": int(done or 0)} for parent_id, total, done in rows}


def _sub_counts_from_children(wo: WorkOrder) -> dict:
    children = list(wo.children)
    return {"total": len(children), "done": sum(1 for child in children if child.status == "done")}


def work_order(wo: WorkOrder, *, sub_counts: dict | None = None) -> dict:
    counts = sub_counts.get(wo.id, {"total": 0, "done": 0}) if sub_counts is not None else _sub_counts_from_children(wo)
    parent = wo.parent if wo.parent_work_order_id else None
    return {
        "id": uid(wo.id),
        "number": wo.number,
        "title": wo.title,
        "description": wo.description,
        "status": wo.status,
        "priority": wo.priority,
        "work_type": wo.work_type,
        "project": project_ref(wo.project),
        "location": location_ref(wo.location),
        "asset": asset_ref(wo.primary_asset),
        "team": team_ref(wo.team),
        "assignees": [user_ref(a.user) for a in wo.assignees if a.user_id and a.user is not None],
        "assignee_teams": [team_ref(a.team) for a in wo.assignees if a.team_id and a.team is not None],
        "watchers": [user_ref(w.user) for w in wo.watchers if w.user is not None],
        "categories": [category_ref(link.category) for link in wo.category_links if link.category is not None],
        "vendor": vendor_ref(wo.vendor),
        "parent_id": uid(wo.parent_work_order_id),
        "parent_number": parent.number if parent is not None else None,
        "parent_completion_policy": wo.parent_completion_policy,
        "sub_work_orders": counts,
        "start_at": iso(wo.start_at),
        "due_at": iso(wo.due_at),
        "completed_at": iso(wo.completed_at),
        "canceled_at": iso(wo.canceled_at),
        "estimated_minutes": wo.estimated_minutes,
        "actual_minutes": int(wo.actual_minutes or 0),
        "is_overdue": bool(wo.is_overdue),
        "is_blocked": bool(wo.is_blocked),
        "budget_code": wo.budget_code,
        "completion_note": wo.completion_note,
        "created_by": user_ref(wo.created_by),
        "created_at": iso(wo.created_at),
        "updated_at": iso(wo.updated_at),
    }


def serialize_many(work_orders: list[WorkOrder]) -> list[dict]:
    counts = sub_counts_for([wo.id for wo in work_orders])
    return [work_order(wo, sub_counts=counts) for wo in work_orders]


def status_history_entry(row) -> dict:
    return {
        "id": uid(row.id),
        "from_status": row.from_status,
        "to_status": row.to_status,
        "changed_by": user_ref(row.changed_by),
        "note": row.note,
        "changed_at": iso(row.changed_at),
    }


def time_entry(row) -> dict:
    return {
        "id": uid(row.id),
        "user": user_ref(row.user),
        "minutes": int(row.minutes),
        "started_at": iso(row.started_at),
        "ended_at": iso(row.ended_at),
        "note": row.note,
        "created_at": iso(row.created_at),
    }


def cost_entry(row) -> dict:
    return {
        "id": uid(row.id),
        "type": row.type,
        "amount": money(row.amount),
        "vendor": vendor_ref(row.vendor),
        "description": row.description,
        "created_at": iso(row.created_at),
    }


def _work_order_stub(wo: WorkOrder) -> dict:
    return {"id": uid(wo.id), "number": wo.number, "title": wo.title, "status": wo.status}


def dependency_entry(dep, *, other: WorkOrder) -> dict:
    return {"id": uid(dep.id), "dependency_type": dep.dependency_type, "work_order": _work_order_stub(other)}


def work_order_detail(wo: WorkOrder) -> dict:
    data = work_order(wo)
    children = list(wo.children)
    data.update(
        {
            "status_history": [status_history_entry(row) for row in wo.status_history],
            "time_entries": [time_entry(row) for row in wo.time_entries],
            "cost_entries": [cost_entry(row) for row in wo.cost_entries],
            "dependencies": {
                "blocked_by": [dependency_entry(dep, other=dep.blocking_work_order) for dep in wo.blocked_by],
                "blocking": [dependency_entry(dep, other=dep.blocked) for dep in wo.blocking],
            },
            "children": serialize_many(children),
            "related_assets": [asset_ref(link.asset) for link in wo.asset_links if link.relationship_type == "related" and link.asset is not None],
        }
    )
    return data

File: apps/ops-web/src/api/contracts/work-orders.ts
import { z } from 'zod'
import { AssetRef, CategoryRef, listOf, LocationRef, PRIORITIES, ProjectRef, TeamRef, UserRef, VendorRef, WORK_TYPES } from './common'

/* Shapes (asme/ops/serializers/work_orders.py) -------------------------------- */

export const SubWorkOrderCounts = z.object({ total: z.number(), done: z.number() })

export const WorkOrder = z.looseObject({
  id: z.string(),
  number: z.number(),
  title: z.string(),
  description: z.string().nullable().optional(),
  status: z.string(),
  priority: z.string(),
  work_type: z.string(),
  project: ProjectRef.nullable().optional(),
  location: LocationRef.nullable().optional(),
  asset: AssetRef.nullable().optional(),
  team: TeamRef.nullable().optional(),
  assignees: z.array(UserRef).default([]),
  assignee_teams: z.array(TeamRef).default([]),
  watchers: z.array(UserRef).default([]),
  categories: z.array(CategoryRef).default([]),
  vendor: VendorRef.nullable().optional(),
  parent_id: z.string().nullable().optional(),
  parent_number: z.number().nullable().optional(),
  parent_completion_policy: z.string().nullable().optional(),
  sub_work_orders: SubWorkOrderCounts.default({ total: 0, done: 0 }),
  start_at: z.string().nullable().optional(),
  due_at: z.string().nullable().optional(),
  completed_at: z.string().nullable().optional(),
  canceled_at: z.string().nullable().optional(),
  estimated_minutes: z.number().nullable().optional(),
  actual_minutes: z.number().default(0),
  is_overdue: z.boolean().default(false),
  is_blocked: z.boolean().default(false),
  budget_code: z.string().nullable().optional(),
  completion_note: z.string().nullable().optional(),
  created_by: UserRef.nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
})
export type WorkOrder = z.infer<typeof WorkOrder>

export const StatusHistoryEntry = z.looseObject({
  id: z.string(),
  from_status: z.string().nullable(),
  to_status: z.string(),
  changed_by: UserRef.nullable().optional(),
  note: z.string().nullable().optional(),
  changed_at: z.string(),
})
export type StatusHistoryEntry = z.infer<typeof StatusHistoryEntry>

export const TimeEntry = z.looseObject({
  id: z.string(),
  user: UserRef.nullable().optional(),
  minutes: z.number(),
  started_at: z.string().nullable().optional(),
  ended_at: z.string().nullable().optional(),
  note: z.string().nullable().optional(),
  created_at: z.string().nullable().optional(),
})
export type TimeEntry = z.infer<typeof TimeEntry>

export const CostEntry = z.looseObject({
  id: z.string(),
  type: z.string(),
  amount: z.number().nullable(),
  vendor: VendorRef.nullable().optional(),
  description: z.string().nullable().optional(),
  created_at: z.string().nullable().optional(),
})
export type CostEntry = z.infer<typeof CostEntry>

export const WorkOrderStub = z.object({ id: z.string(), number: z.number(), title: z.string(), status: z.string() })
export type WorkOrderStub = z.infer<typeof WorkOrderStub>

export const DependencyEntry = z.looseObject({
  id: z.string(),
  dependency_type: z.string().optional(),
  work_order: WorkOrderStub,
})
export type DependencyEntry = z.infer<typeof DependencyEntry>

export const WorkOrderDetail = WorkOrder.extend({
  status_history: z.array(StatusHistoryEntry).default([]),
  time_entries: z.array(TimeEntry).default([]),
  cost_entries: z.array(CostEntry).default([]),
  dependencies: z.object({ blocked_by: z.array(DependencyEntry).default([]), blocking: z.array(DependencyEntry).default([]) }).default({ blocked_by: [], blocking: [] }),
  children: z.array(WorkOrder).default([]),
  related_assets: z.array(AssetRef).default([]),
})
export type WorkOrderDetail = z.infer<typeof WorkOrderDetail>

export const TabCounts = z.object({ todo: z.number(), done: z.number() })
export type TabCounts = z.infer<typeof TabCounts>

export const WorkOrderList = listOf(WorkOrder).extend({ tabs: TabCounts.optional() })
export type WorkOrderList = z.infer<typeof WorkOrderList>

export const CompleteResponse = z.object({ work_order: WorkOrderDetail, follow_up: WorkOrder.nullable().optional() })
export type CompleteResponse = z.infer<typeof CompleteResponse>

/* List parameters (services/work_orders.py LIST_FILTERS / LIST_SORTS) --------- */

export const WORK_ORDER_FILTER_NAMES = ['status', 'priority', 'work_type', 'project', 'location', 'asset', 'team', 'assignee', 'category', 'due', 'created_by', 'parent', 'vendor'] as const
export type WorkOrderFilterName = (typeof WORK_ORDER_FILTER_NAMES)[number]
export type WorkOrderFilters = Partial<Record<WorkOrderFilterName, string[]>>

export type WorkOrderTab = 'todo' | 'done' | 'all'
export type WorkOrderSort = 'priority' | '-priority' | 'due_at' | '-due_at' | 'updated_at' | '-updated_at' | 'created_at' | '-created_at' | 'number' | '-number' | 'title' | '-title'
export const DEFAULT_WORK_ORDER_SORT: WorkOrderSort = '-updated_at'

export const WORK_ORDER_SORT_OPTIONS: Array<{ value: WorkOrderSort; label: string }> = [
  { value: '-priority', label: 'Priority: Highest First' },
  { value: 'due_at', label: 'Due Date: Soonest First' },
  { value: '-updated_at', label: 'Last Updated' },
  { value: '-created_at', label: 'Created Date' },
  { value: 'number', label: 'Number' },
  { value: 'title', label: 'Title' },
]

export const DUE_FILTER_OPTIONS = [
  { value: 'overdue', label: 'Overdue' },
  { value: 'today', label: 'Today' },
  { value: 'week', label: 'This week' },
  { value: 'month', label: 'This month' },
  { value: 'none', label: 'No due date' },
]

export const COST_TYPES = ['parts', 'labor', 'vendor', 'other'] as const
export type CostType = (typeof COST_TYPES)[number]
export const COST_TYPE_LABELS: Record<CostType, string> = { parts: 'Parts', labor: 'Labor', vendor: 'Vendor', other: 'Other' }

export const ASSET_STATUSES = ['online', 'offline_planned', 'offline_unplanned', 'do_not_track', 'retired'] as const
export const DOWNTIME_TYPES = ['planned', 'unplanned'] as const

export interface WorkOrderListParams {
  q?: string
  tab?: WorkOrderTab
  sort?: WorkOrderSort | string
  filters?: WorkOrderFilters
  limit?: number
}

/* Request bodies (services/work_orders.py CREATE_SPEC and friends) ------------ */

export interface WorkOrderInput {
  title: string
  description?: string | null
  priority?: string
  work_type?: string
  project_id?: string | null
  location_id?: string | null
  primary_asset_id?: string | null
  team_id?: string | null
  vendor_id?: string | null
  parent_id?: string | null
  start_at?: string | null
  due_at?: string | null
  estimated_minutes?: number | null
  budget_code?: string | null
  parent_completion_policy?: 'manual' | 'auto'
  assignee_user_ids?: number[]
  assignee_team_ids?: string[]
  watcher_user_ids?: number[]
  category_ids?: string[]
  asset_ids?: string[]
  draft?: boolean
  /** Only meaningful on PATCH: publishes a draft. */
  status?: 'open'
}

export interface TimeEntryInput {
  minutes?: number | null
  started_at?: string | null
  ended_at?: string | null
  note?: string | null
  user_id?: number
}

export interface CostEntryInput {
  type: CostType | string
  amount: number
  vendor_id?: string | null
  description?: string | null
}

export interface CompleteInput {
  note?: string | null
  time_entries?: TimeEntryInput[]
  cost_entries?: CostEntryInput[]
  asset_status?: { status: string; downtime_type?: string | null; downtime_reason?: string | null; note?: string | null } | null
  follow_up?: { title: string; description?: string | null; priority?: string; work_type?: string; due_at?: string | null } | null
}

export type TransitionAction = 'start' | 'hold' | 'resume' | 'cancel' | 'reopen'

/* Form schema (create / edit pane) ------------------------------------------- */

const wholeNumber = z.string().regex(/^\d*$/, 'Use whole numbers.')

export const WorkOrderFormValues = z.object({
  title: z.string().trim().min(1, 'Give the work order a title.').max(240, 'Keep the title under 240 characters.'),
  description: z.string().max(20000, 'Keep the description under 20,000 characters.'),
  sub_work_orders: z.array(z.object({ title: z.string().trim().max(240, 'Keep the title under 240 characters.') })),
  project_id: z.string().nullable(),
  location_id: z.string().nullable(),
  primary_asset_id: z.string().nullable(),
  assignee_user_ids: z.array(z.number()),
  team_id: z.string().nullable(),
  estimated_hours: wholeNumber,
  estimated_minutes: wholeNumber,
  due_at: z.string(),
  start_at: z.string(),
  work_type: z.enum(WORK_TYPES),
  priority: z.enum(PRIORITIES),
  category_ids: z.array(z.string()),
  vendor_id: z.string().nullable(),
  budget_code: z.string().trim().max(60, 'Keep the budget code under 60 characters.'),
  watcher_user_ids: z.array(z.number()),
  parent_id: z.string().nullable(),
})
export type WorkOrderFormValues = z.infer<typeof WorkOrderFormValues>

export const EMPTY_WORK_ORDER_FORM: WorkOrderFormValues = {
  title: '',
  description: '',
  sub_work_orders: [],
  project_id: null,
  location_id: null,
  primary_asset_id: null,
  assignee_user_ids: [],
  team_id: null,
  estimated_hours: '',
  estimated_minutes: '',
  due_at: '',
  start_at: '',
  work_type: 'reactive',
  priority: 'none',
  category_ids: [],
  vendor_id: null,
  budget_code: '',
  watcher_user_ids: [],
  parent_id: null,
}

/** Combined estimate in minutes, or null when both inputs are blank. */
export function estimateToMinutes(hours: string, minutes: string): number | null {
  const h = hours.trim() === '' ? 0 : Number(hours)
  const m = minutes.trim() === '' ? 0 : Number(minutes)
  if (hours.trim() === '' && minutes.trim() === '') return null
  if (!Number.isFinite(h) || !Number.isFinite(m)) return null
  return h * 60 + m
}

export function minutesToEstimate(total: number | null | undefined): { estimated_hours: string; estimated_minutes: string } {
  if (total === null || total === undefined) return { estimated_hours: '', estimated_minutes: '' }
  return { estimated_hours: String(Math.floor(total / 60)), estimated_minutes: String(total % 60) }
}
